utils/diffing.mli

```diff
@@ -59,71 +59,97 @@
-
-(** [diff ~weight ~test ~update state l r] computes
-    the diff between [l] and [r], using the initial state [state].
-    - [test st xl xr] tests if the elements [xl] and [xr] are
-      compatible ([Ok]) or not ([Error]).
-    - [weight ch] returns the weight of the change [ch].
-      Used to find the smallest patch.
-    - [update ch st] returns the new state after applying a change.
-*)
-val diff :
-  weight:(('l, 'r, 'eq, 'diff) change -> int) ->
-  test:('state -> 'l -> 'r -> ('eq, 'diff) result) ->
-  update:(('l, 'r, 'eq, 'diff) change -> 'state -> 'state) ->
-  'state -> 'l array -> 'r array -> ('l, 'r, 'eq, 'diff) patch
-
-(** {1 Variadic diffing}
-
-    Variadic diffing allows to expand the lists being diffed during diffing.
-*)
-
-type ('l, 'r, 'e, 'd, 'state) update =
-  | Without_extensions of (('l,'r,'e,'d) change -> 'state -> 'state)
-  | With_left_extensions of
-      (('l,'r,'e,'d) change -> 'state -> 'state * 'l array)
-  | With_right_extensions of
-      (('l,'r,'e,'d) change -> 'state -> 'state * 'r array)
-
-(** [variadic_diff ~weight ~test ~update state l r] behaves as [diff]
-    with the following difference:
-    - [update] must now be an {!update} which indicates in which direction
-      the expansion takes place.
-*)
-val variadic_diff :
-  weight:(('l, 'r, 'eq, 'diff) change -> int) ->
-  test:('state -> 'l -> 'r -> ('eq, 'diff) result) ->
-  update:('l, 'r, 'eq, 'diff, 'state) update ->
-  'state -> 'l array -> 'r array -> ('l, 'r, 'eq, 'diff) patch
-
-
-val default_weight : _ change -> int
-
-(** Printing default function *)
+(** The core types of a diffing implementation *)
+module type Defs = sig
+  type left
+  type right
+  type eq
+  (** Detailled equality trace *)
+
+  type diff
+  (** Detailled difference trace *)
+
+  type state
+  (** environment of a partial patch *)
+end
+
+(** The kind of changes which is used to share printing and styling
+    across implementation*)
 type change_kind =
   | Deletion
   | Insertion
   | Modification
   | Preservation
-
-val classify: _ change -> change_kind
 val prefix: Format.formatter -> (int * change_kind) -> unit
 val style: change_kind -> Misc.Color.style list
 
-(** Enriched analysis in presence of keys *)
+
+module Generic: sig
+  type ('left,'right,'eq,'diff) change =
+    | Delete of 'left
+    | Insert of 'right
+    | Keep of 'left * 'right *' eq
+    | Change of 'left * 'right * 'diff
+   val classify: _ change -> change_kind
+end
+
+(** [Define(Defs)] creates the diffing types from the types
+    defined in [Defs] and the functors that need to be instantatied
+    with the diffing algorithm parameters
+*)
+module Define(D:Defs): sig
+  open D
+
+  (** The type of potential changes on a list. *)
+  type change =
+    | Delete of left
+    | Insert of right
+    | Keep of left * right * eq
+    | Change of left * right * diff
+  val generalize: change -> (left,right,eq,diff) Generic.change
```

Comment:
I still don't understand why we are doing this instead of `type change = (left,right,eq,diff) Generic.change`.

I asked @Octachron in https://github.com/Octachron/ocaml/pull/3#issuecomment-858054701, and his answer was

> This has the advantage of making the update and test handles the correct change type by construction since the specialized change type is the only one in scope.

I'm not sure what this means (it's probably about typed-based constructor disambiguation?), but I don't really know what explanation would be convincing enough to justify a second copy of a type and a weird `generalize` function in the API.

Reply:
I intended to say that this duplicated definition reduces the need for type annotations in implementation because type-directed disambiguation is not needed. However, looking at the code of the existing implementations, I use the annotation half of the time (and it avoids an open). My conclusion is thus that this was a premature API optimization.